src/parser.ts

```diff
@@ -1,4 +1,151 @@
+      return { doc: v1Doc, errors }
+    default:
+      return {
+        errors: [
+          new ValidationError(`version property not valid: ${doc.version}`, "#/version")
+        ]
+      }
+  }
+}
+
+export function isV0Schema(schema?: VUnknownSchema): schema is V0Schema {
+  return schema?.version === 'v0';
+}
+
+export function isV1Schema(schema?: VUnknownSchema): schema is V1Schema {
+  return schema?.version === 'v1-draft';
+}
+
+/**
+ * Validates a V1 document.
+ */
+class V1Validator {
+  errors: ValidationError[]
+  location: string[]
+  doc: any
+
+  constructor(doc: V1Schema) {
+    this.doc = doc as any
+    this.errors = []
+    this.location = ['#']
+  }
+
+  /**
+   * Validate the document and return any errors
+   */
+  validate(): ValidationError[] {
+    this.errors = []
+    this.validateNode(this.doc)
+    return this.errors
+  }
+
+  /**
+   * Recursively walk through the untyped document and validate each node.
+   * This saves us a lot of code but might be a little bit slower.
+   */
+  validateNode(node: any) {
+    this.validateTypedInterface(node)
+
+    if (node && typeof node === 'object') {
+      // i don't think we need to validate array children
+      if (Array.isArray(node)) return
+
+      for (const key in node) {
+        if (Object.prototype.hasOwnProperty.call(node, key)) {
+          const child = node[key]
+          if (typeof child === 'object') {
+            this.location.push(key)
+            this.validateNode(child);
+            this.location.pop()
+          }
+        }
+      }
+    }
+  }
+
+  recordError(msg: string) {
+    this.errors.push(
+      new ValidationError(msg, this.getLocation())
+    )
+  }
+
+  /**
+   * Validates that a node conforms to the rules of
+   * the XtpTyped interface. Validates what we can't
+   * catch in JSON Schema validation.
+   */
+  validateTypedInterface(prop?: XtpTyped): void {
+    if (!prop) return
+
+    const validTypes = ['string', 'number', 'integer', 'boolean', 'object', 'array', 'buffer'];
+    if (prop.type && !validTypes.includes(prop.type)) {
+      this.recordError(`Invalid type '${prop.type}'. Options are: ${validTypes.map(t => `'${t}'`).join(', ')}`)
+    }
+
+    if (prop.format) {
+      let validFormats: XtpFormat[] = [];
+      if (prop.type === 'string') {
+        validFormats = ['date-time', 'byte'];
+      } else if (prop.type === 'number') {
+        validFormats = ['float', 'double'];
+      } else if (prop.type === 'integer') {
+        validFormats = ['int32', 'int64'];
+      }
+
+      if (!validFormats.includes(prop.format)) {
+        this.recordError(`Invalid format ${prop.format} for type ${prop.type}. Valid formats are: ${validFormats.join(', ')}`)
+      }
+    }
+
+
+    if (prop.items) this.validateTypedInterface(prop.items)
+
+    if (prop.additionalProperties) {
+      this.validateTypedInterface(prop.additionalProperties)
+
+      // here we are adding some extra constraints on the value type
```

Comment:
Just above this we validate that additionalProperties meets XtpTyped rules. but here we are adding some extra constraints because we don't want to support recursive value types here just yet. We will likely do the same for array. but this code can be deleted once I feel comfortable that the PDKs can handle it.

Reply:
A ha – so to confirm my understanding, is it the trickiness of writing recursive casting code that's steering us away from supporting recursive types?

Reply:
I ended up backing this change out. Though if it seems too complex in other bindgens then maybe we put it back.